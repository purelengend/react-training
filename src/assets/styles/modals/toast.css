.toast-container {
  width: 250px;
  align-items: center;
  position: fixed;
  top: 3%;
  left: 50%;
  transform: translateX(-50%);
  border-radius: var(--border-primary);
  padding: 13px 11px;
  z-index: 99;
  opacity: 0;
  transition: all 0.3s ease-in-out;
}

.success {
  background-color: var(--toast-bg);
}

.error {
  background-color: var(--warning-color);
}

.toast-message-wrapper {
  color: var(--text-primary);
  font-weight: var(--font-weight-bold);
  line-height: var(--line-height-lg);
  gap: 8px;
}

.error-icon {
  padding-top: 4px;
}
