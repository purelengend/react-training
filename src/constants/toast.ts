export const TOAST_TIME = 2500;

export const TOAST_MSG = {
  ADD: 'Add successfully!',
  DELETE: 'Delete successfully!',
  EDIT: 'Edit successfully!',
  ERROR: 'Request failed: Something went wrong.'
} as const;
