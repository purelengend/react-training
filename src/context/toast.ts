import { createContext } from 'react';

export interface ToastContextProps {
  toast: {
    message: string;
    isSuccess: boolean;
    isVisible: boolean;
  };
  showToast: (message: string, isSuccess: boolean) => void;
  hideToast: () => void;
}

export const ToastContext = createContext<ToastContextProps>({
  toast: {
    message: '',
    isSuccess: true,
    isVisible: false
  },
  showToast() {},
  hideToast() {}
});
