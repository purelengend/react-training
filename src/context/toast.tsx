import useToast from '@hooks/useToast';
import { ReactNode, createContext, memo, useMemo } from 'react';
import isEqual from 'react-fast-compare';

export interface ToastContextProps {
  toast: {
    message: string;
    isSuccess: boolean;
    isVisible: boolean;
  };
  showToast: (message: string, isSuccess: boolean) => void;
  hideToast: () => void;
}

export const ToastContext = createContext<ToastContextProps>({
  toast: {
    message: '',
    isSuccess: true,
    isVisible: false
  },
  showToast() {},
  hideToast() {}
});

interface ToastContextProviderProps {
  children: ReactNode;
}

export const ToastContextProvider = memo(
  ({ children }: ToastContextProviderProps) => {
    const { toast, showToast, hideToast } = useToast();
    const toastContextValue = useMemo(
      () => ({
        toast,
        showToast,
        hideToast
      }),
      [toast, showToast, hideToast]
    );
    return (
      <ToastContext.Provider value={toastContextValue}>
        {children}
      </ToastContext.Provider>
    );
  },
  isEqual
);

ToastContextProvider.whyDidYouRender = true;
