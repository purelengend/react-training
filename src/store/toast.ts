export enum ToastActionKind {
  Toast = 'TOAST'
}

export interface ToastAction {
  type: ToastActionKind;
  payload: {
    message?: string;
    isSuccess?: boolean;
    isVisible: boolean;
  };
}

export interface ToastState {
  message: string;
  isSuccess: boolean;
  isVisible: boolean;
}

export const initialToastState: ToastState = {
  message: '',
  isSuccess: true,
  isVisible: false
};

export const toastReducer = (
  state: ToastState,
  action: ToastAction
): ToastState => {
  const { type, payload } = action;
  switch (type) {
    case ToastActionKind.Toast:
      return {
        ...state,
        message: payload.message ?? state.message,
        isSuccess: payload.isSuccess ?? state.isSuccess,
        isVisible: payload.isVisible
      };
    default:
      return state;
  }
};
