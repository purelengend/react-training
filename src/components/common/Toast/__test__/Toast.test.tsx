import { renderer } from '@src/test/test-utils';
import { Toast } from '..';
import { ToastType } from '@store/toast';

describe('Toast test case', () => {
  const mockMessage = 'mocked message';
  let mockType = 'SUCCESS' as ToastType;
  it('should render correctly and be hidden by default', () => {
    const mockToast = renderer(
      <Toast message={mockMessage} toastType={mockType} />
    );

    expect(mockToast.toJSON()).toMatchSnapshot();
  });

  it('should be visible', () => {
    const mockToast = renderer(
      <Toast message={mockMessage} toastType={mockType} isVisible={true} />
    );

    expect(mockToast.toJSON()).toMatchSnapshot();
  });

  mockType = 'ERROR' as ToastType;
  it('should render error toast when toast type is ERROR', () => {
    const mockToast = renderer(
      <Toast message={mockMessage} toastType={mockType} />
    );

    expect(mockToast.toJSON()).toMatchSnapshot();
  });
});
