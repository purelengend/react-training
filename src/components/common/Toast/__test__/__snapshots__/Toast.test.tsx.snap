// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toast test case > should be visible 1`] = `
<div
  className="d-flex _toast-container_070444 _error_070444"
  style={
    {
      "opacity": "1",
      "visibility": "visible",
    }
  }
>
  <div
    className="d-flex _toast-message-wrapper_070444"
  >
    <img
      alt="Check Mark Icon"
      className="primary-icon _error-icon_070444"
      src="/src/assets/icons/cross-icon.svg"
    />
    <p>
      mocked message
    </p>
  </div>
</div>
`;

exports[`Toast test case > should render correctly and be hidden by default 1`] = `
<div
  className="d-flex _toast-container_070444 _error_070444"
  style={
    {
      "opacity": "0",
      "visibility": "hidden",
    }
  }
>
  <div
    className="d-flex _toast-message-wrapper_070444"
  >
    <img
      alt="Check Mark Icon"
      className="primary-icon _error-icon_070444"
      src="/src/assets/icons/cross-icon.svg"
    />
    <p>
      mocked message
    </p>
  </div>
</div>
`;

exports[`Toast test case > should render error toast when toast type is ERROR 1`] = `
<div
  className="d-flex _toast-container_070444 _error_070444"
  style={
    {
      "opacity": "0",
      "visibility": "hidden",
    }
  }
>
  <div
    className="d-flex _toast-message-wrapper_070444"
  >
    <img
      alt="Check Mark Icon"
      className="primary-icon _error-icon_070444"
      src="/src/assets/icons/cross-icon.svg"
    />
    <p>
      mocked message
    </p>
  </div>
</div>
`;
